### PR TITLE
Develop Offline Tuning

### PR DESCRIPTION
**Pull request by Jay0521**

### CODE REVIEW

File: library/src/amd_detail/hipblaslt.cpp [hunk 1]
```diff
@@ -34,12 +34,42 @@
...
+        std::cerr << "override file git version: " << file_version << std::endl;
+        if (file_version == git_version)
+            return true;
+    }
+
+    overridePath = "";
+
+    std::cerr << "The hipBLASLt git version and the override file git version are not the same." << std::endl;
+
+    setenv("HIPBLASLT_TUNING_OVERRIDE_FILE", overridePath.c_str(), 1);
```
Comment (jichangjichang) [inline library/src/amd_detail/hipblaslt.cpp]: why do you do this

Comment (Jay0521) [reply library/src/amd_detail/hipblaslt.cpp]: The file won’t be able to be opened later during execution if this check fails here
However, it looks like I can refer to the code related to get_logger_layer_mode() to change the implementation
 I will proceed with the modifications

File: docs/how-to-use-user-offline-tuning.rst [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+.. _how-to-use-hipblaslt-offline-tuning:
+
+********************************
+User Offline Tuning
+********************************
+
+hipblaslt-bench can be used to find the best performing GEMM kernel for each of a given set of GEMM problems.
+
+It has a command line interface. (See :ref:`clients` for more details.)
```
Comment (imcarsonliao) [inline docs/how-to-use-user-offline-tuning.rst:13]: The hyperlink points to this page itself.

File: docs/how-to-use-user-offline-tuning.rst [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+The result of best solution index for a given problem size can be directly used in future GEMM calls.
+
+However, please note that these indices cannot be reused across library releases or across different device architectures.
+
+Running
+=================================
+
+Follow these steps:
+
+1. To generate the hipblaslt-bench, logging can be used by setting environment variable ``HIPBLASLT_LOG_MASL=32``. (see :ref:`Logging and heuristics <logging-heuristics>` for more details.)
```
Comment (imcarsonliao) [inline docs/how-to-use-user-offline-tuning.rst:24]: typo: HIPBLASLT_LOG_MASK=32.
rephrase in this way: Generate tuning command line by setting...

File: docs/how-to-use-user-offline-tuning.rst [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+2. User can use ``hipblaslt-bench`` to find and store the best solution indices for those GEMM problems, by setting the environment variable ``HIPBLASLT_TUNING_FILE=<file_name>``, where ``<file_name>`` points to the tuning file.
```
Comment (imcarsonliao) [inline docs/how-to-use-user-offline-tuning.rst:25]: rephrase in this way: Set the env variable .. ... to tune and store the tuning result...

File: docs/how-to-use-user-offline-tuning.rst [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+- Please note that the default settings will have the following changes in the tuning environment.
+.. code-block:: bash
+
+    --iters |-i <value>             Iterations to run inside timing loop                                                (Default value is: 1000)
+    --cold_iters |-j <value>        Cold Iterations to run before entering the timing loop                              (Default value is: 1000)
+    --algo_method <value>           Use different algorithm search API. Options: heuristic, all, index (still considered all).                  (Default value is: all)
+    --requested_solution <value>    Requested solution num. Set to -1 to get all solutions. Only valid when algo_method is set to heuristic.    (Default value is: -1)
+    --rotating <value>              Use rotating memory blocks for each iteration, size in MB.                          (Default value is: 512)
+- After the tuning completes, the expected output is as follows:
+.. code-block:: bash
```
Comment (imcarsonliao) [inline docs/how-to-use-user-offline-tuning.rst:35]: "code-block:: bash" accidently shows on the page

File: docs/how-to-use-user-offline-tuning.rst [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+
+    ./hipblaslt-bench --api_method c -m 64 -n 64 -k 1024 --lda 64 --ldb 1024 --ldc 64 --ldd 64  --stride_a 0 --stride_b 0 --stride_c 0 --stride_d 0  --alpha 1.100000 --beta 0.900000 --transA N --transB N --batch_count 1  --a_type f32_r --b_type f32_r --c_type f32_r --d_type f32_r --scale_type f32_r --bias_type f32_r   --compute_type f32_r
+    Winner: transA,transB,grouped_gemm,batch_count,m,n,k,alpha,lda,stride_a,beta,ldb,stride_b,ldc,stride_c,ldd,stride_d,a_type,b_type,c_type,d_type,compute_type,scaleA,scaleB,scaleC,scaleD,amaxD,activation_type,bias_vector,bias_type,rotating_buffer,hipblaslt-Gflops,hipblaslt-GB/s,us,soulution_index
+    N,N,0,1,64,64,1024,1.1,64,65536,0.9,1024,65536,64,4096,64,4096,f32_r,f32_r,f32_r,f32_r,f32_r,0,0,0,0,0,none,0,f32_r,512,903.229,55.8608,9.28735,49257
+
+
+3. If the output is stored in a file, these results can be used to override default kernel selection with the kernels found, by setting the environment variable ``HIPBLASLT_TUNING_OVERRIDE_FILE=<file_name>``, where ``<file_name>`` points to the override file.
```
Comment (imcarsonliao) [inline docs/how-to-use-user-offline-tuning.rst:42]: rephrase: set the env var... to load the tuning file...

File: docs/how-to-use-user-offline-tuning.rst [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+- For example, user can use hisblaslt-bench (get heuristic) to obtain solutions for a problem, which will include the best tuning solution index.
+.. code-block:: bash
```
Comment (imcarsonliao) [inline docs/how-to-use-user-offline-tuning.rst:44]: again, "code-block:: bash"

File: docs/how-to-use-user-offline-tuning.rst (hunk 3 above)
Comment (imcarsonliao) [inline docs/how-to-use-user-offline-tuning.rst:24]: link points to itself

**Pull request closed without merging**